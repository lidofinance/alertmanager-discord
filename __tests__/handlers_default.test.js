const axios = require("axios");
jest.mock("axios");

const { handleHook, handleHealthcheck, getMentions, getFields } = require("../handlers_default");

function getAlert(i) {
  return {
    status: "resolved",
    labels: { alertname: "activate" },
    annotations: {
      description: "cinema",
      summary: "donor" + String(i).padStart(2, "0"),
    },
  };
}

test("hook works (no mentions)", async () => {
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    logger: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    params: { slug: "test" },
    query: { q: "42" },

    request: {
      body: {
        alerts: Array.from({ length: 11 }, (_, i) => getAlert(i)),
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx, () => {});

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(11);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("hook works (missing annotations subfield)", async () => {
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    logger: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    params: { slug: "test" },
    query: {},

    request: {
      body: {
          alerts: [
              {
                status: "resolved",
                labels: { alertname: "activate" },
                annotations: {
                  summary: "here was no description"
                },
              },
              {
                status: "resolved",
                labels: { alertname: "activate" },
                annotations: {
                  description: "here was no summary"
                },
              },
          ],
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx, () => {});

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(2);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("hook works (mentions)", async () => {
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    logger: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    params: { slug: "test" },
    query: {},

    request: {
      body: {
        alerts: Array.from({ length: 11 }, (_, i) => {
          const body = getAlert(i);
          body.labels.mentions = `user${i},42,`;
          return body;
        }),
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx, () => {});

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(11);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("hook works (fields)", async () => {
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    logger: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    params: { slug: "test" },
    query: {},

    request: {
      body: {
        alerts: [
          {
            status: "resolved",
            labels: { alertname: "activate" },
            annotations: {
              summary: "Obi-Wan Kenobi says",
              inline_fields: ["- **hello**", "- there"].join("\n\n"),
            },
          },
        ],
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx, () => { });

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("healthcheck works", async () => {
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    logger: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
    params: { slug: "test" },
    query: {},
  };

  axios.get.mockResolvedValue(null);

  await handleHealthcheck(ctx, () => {});

  expect(ctx.status).toBe(200);
  expect(ctx.body.uptime).toBeDefined();
  expect(axios.get.mock.calls).toMatchSnapshot();
});

test("getMentions works (no label)", () => {
  const mentions = getMentions({
    labels: {},
  });

  expect(mentions.length).toBe(0);
});

test("getMentions works (valid label's value)", () => {
  const mentions = getMentions({
    labels: {
      mentions: "123,456,",
    },
  });

  expect(mentions).toStrictEqual(["<@123>", "<@456>"]);
});

test("getFields works (no inline_fields)", () => {
  const fields = getFields({
    annotations: {},
  });

  expect(fields.length).toBe(0);
});

test("getFields works (not a list)", () => {
  const fields = getFields({
    annotations: {
      inline_fields: "**hm**",
    },
  });

  expect(fields.length).toBe(0);
})

test("getFields works (valid list)", () => {
  const fields = getFields({
    annotations: {
      inline_fields: "- **hm**",
    },
  });

  expect(fields.length).toBe(1);
})

test("getFields works (nested list)", () => {
  const fields = getFields({
    annotations: {
      inline_fields: "- **hm**\n\t+ **hmmm**",
    },
  });

  expect(fields.length).toBe(1);
})
