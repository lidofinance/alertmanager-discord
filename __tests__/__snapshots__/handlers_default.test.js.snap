// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`healthcheck works 1`] = `
Array [
  Array [
    "/dev/null",
  ],
]
`;

exports[`hook works (fields) 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": undefined,
          "fields": Array [
            Object {
              "inline": true,
              "name": "",
              "value": "**hello**",
            },
            Object {
              "inline": true,
              "name": "",
              "value": "there",
            },
          ],
          "title": "Obi-Wan Kenobi says",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
]
`;

exports[`hook works (mentions) 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user0> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor00",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user1> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor01",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user2> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor02",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user3> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor03",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user4> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor04",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user5> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor05",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user6> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor06",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user7> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor07",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user8> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor08",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user9> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor09",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@user10> <@42>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor10",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
]
`;

exports[`hook works (missing annotations subfield) 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": undefined,
          "title": "here was no description",
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "here was no summary",
          "title": undefined,
        },
      ],
    },
    Object {
      "params": Object {},
    },
  ],
]
`;

exports[`hook works (no mentions) 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor00",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor01",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor02",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor03",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor04",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor05",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor06",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor07",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor08",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor09",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "title": "donor10",
        },
      ],
    },
    Object {
      "params": Object {
        "q": "42",
      },
    },
  ],
]
`;
